### PR TITLE
Remove clone() when removing extra stack operations

### PR DESCRIPTION
**Pull request by CohenArthur**
Replace the double `clone()` in `remove_unused_stack_addr_and_stack_load()` with a `while let` expression.

### CODE REVIEW

File: src/optimize/stack2reg.rs [hunk 1]
```diff
@@ -314,19 +312,19 @@ fn remove_unused_stack_addr_and_stack_load(opt_ctx: &mut OptimizeContext<'_>) {
...
-
-        for &inst in stack_slot_users.stack_load.clone().iter() {
-            if stack_addr_load_insts_users.get(&inst).map(|users| users.is_empty()).unwrap_or(true) {
-                stack_slot_users.remove_unused_load(&mut opt_ctx.ctx.func, inst);
-            }
-        }
+        stack_slot_users
+            .stack_addr
+            .iter()
+            .filter(|inst| stack_addr_load_insts_users.get(inst).map(|users| users.is_empty()).unwrap_or(true))
```
[the remark this answers is not in the document]
Comment (CohenArthur) [reply src/optimize/stack2reg.rs]: What I mean is that `drain_filter()` isn't implemented for `HashSet`. See [this](https://github.com/rust-lang/rust/issues/59618). It's implemented in the hashbrown crate, but from my understanding it hasn't landed yet.

[the remark this answers is not in the document]
Comment (bjorn3) [reply src/optimize/stack2reg.rs]: Ah, I thought it was a `Vec`. In that case I don't know how to do it.

[the remark this answers is not in the document]
Comment (CohenArthur) [reply src/optimize/stack2reg.rs]: I'm going to look into it. Either we wait for it to land, or we implement our own as a small patch for now.

[the remark this answers is not in the document]
Comment (bjorn3) [reply src/optimize/stack2reg.rs]: I don't think it is possible to implement `drain_filter` without access to internal fields of the `HashSet` implementation. Maybe directly use `hashbrown::HashSet`? (With `rustc_data_structures::fx::FxHasher` as hasher)

[the remark this answers is not in the document]
Comment (CohenArthur) [reply src/optimize/stack2reg.rs]: Okay, done. However, the behavior of `drain_filter()` in hashbrown hashset is gonna be reverted. For now, I simply had to revert the predicate, but once hashbrown gets updated (0.8.2) we'll have to reverse it. So far it's only a comment in the code, and I specified the hashbrown version as "0.8.1"

[the remark this answers is not in the document]
Comment (CohenArthur) [reply src/optimize/stack2reg.rs]: Here is the changelog explaining the behavior of `drain_filter()`: https://github.com/rust-lang/hashbrown/blob/HEAD/CHANGELOG.md

**Review by bjorn3 — APPROVED** [review]
Thanks!

**Pull request merged**